# eBay Car Sales 

## Introduction 

We will be working on a dataset of used cars from *eBay Kleinanzeigen*, a [classifieds](https://en.wikipedia.org/wiki/Classified_advertising) section of the German eBay website.

The dataset was originally scraped and uploaded to [Kaggle](https://www.kaggle.com/orgesleka/used-cars-database/data).  The version of the dataset we are working with is a sample of 50,000 data points that was prepared by [Dataquest](https://www.dataquest.io) including simulating a less-cleaned version of the data.

The data dictionary provided with data is as follows:

- `dateCrawled` - When this ad was first crawled. All field-values are taken from this date.
- `name` - Name of the car.
- `seller` - Whether the seller is private or a dealer.
- `offerType` - The type of listing
- `price` - The price on the ad to sell the car.
- `abtest` - Whether the listing is included in an A/B test.
- `vehicleType` - The vehicle Type.
- `yearOfRegistration` - The year in which which year the car was first registered.
- `gearbox` - The transmission type.
- `powerPS` - The power of the car in PS.
- `model` - The car model name.
- `kilometer` - How many kilometers the car has driven.
- `monthOfRegistration` - The month in which which year the car was first registered.
- `fuelType` - What type of fuel the car uses.
- `brand` - The brand of the car.
- `notRepairedDamage` - If the car has a damage which is not yet repaired.
- `dateCreated` - The date on which the eBay listing was created.
- `nrOfPictures` - The number of pictures in the ad.
- `postalCode` - The postal code for the location of the vehicle.
- `lastSeenOnline` - When the crawler saw this ad last online.


The aim of this project is to clean the data and analyze the included used car listings.

## Import the working libraries

In [1]:
import numpy as np
import pandas as pd

## Import and clean the dataset

In [2]:
autos = pd.read_csv('autos.csv', encoding = 'latin1')                  

In [3]:
autos.info()
#autos.head(5)
autos

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

**My first observations are:**
- Dataset contains 20 columns most of which are strings;
- Data isn't sorted based on any principle;
- There are some null entries for the 'vehicleType', 'gearbox', 'model', 'fuelType', 'notRepairedDamage';
- Some of the columns could be easily converted from object dtype to more workable dtypes, e.g. 'notRepairedDamage' could be converted in a series of bool;
- The column labels are CamleCase so we might want them in snake_case for convenince and consistency;
- There are already some int64 dtype columns like 'monthOfRegistration', 'powerPS';
- The 'price', 'odometer' could be converted in numerical series and the column labels renamed to be explicit;

**Converting The Column Labels Into snake_case**

In [4]:
#The camel case columns before change
print(autos.columns)

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


In [5]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen']

#print(autos.info())

**Columns investigation**
- Which column would I drop? ('seller' as we can simply specify that all are private sellers, 'name' could easily be replaced with 'brand' followed by the 'model' series) 
- Columns that need more investigation? ('nr_of_pictures' to see if it can be dropped), 
- Examples of numeric series stored as object? ('price' and 'odometer' could be cleaned of the measuring units and the labels could be renamed)

**Clean the 'price' and 'odometer' columns and convert into numeric series**

In [6]:
print(list(autos.columns))
if {"price","odometer"}.issubset(set(autos.columns)) and autos["price"].dtype == object and autos["odometer"].dtype == object:
    clean_price = autos["price"].str.replace("$","").str.replace(",","").astype(float)
    clean_odometer = autos["odometer"].str.replace("km", "").str.replace(",","").astype(float)
    autos["price"] = clean_price
    autos["odometer"] = clean_odometer
    autos.rename({"odometer":"odometer_km", "price":"price_dollars"}, axis=1, inplace=True)

autos.info()


['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest', 'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model', 'odometer', 'registration_month', 'fuel_type', 'brand', 'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code', 'last_seen']
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   date_crawled        50000 non-null  object 
 1   name                50000 non-null  object 
 2   seller              50000 non-null  object 
 3   offer_type          50000 non-null  object 
 4   price_dollars       50000 non-null  float64
 5   abtest              50000 non-null  object 
 6   vehicle_type        44905 non-null  object 
 7   registration_year   50000 non-null  int64  
 8   gearbox             47320 non-null  object 
 9   power_ps            50000 non-null  int64  
 10  model               4724

**We can see that the 'offer_type' has one "Gesuch" which translates as "Request". This entry is actually not a car being on sale but somebody wanting to buy this specific car which could affect the accuracy of the calculations. It has to be removed. After removal we could also remove the offer_type column since it is implicit.**

In [7]:
autos["offer_type"].value_counts()
autos = autos[autos["offer_type"] != "Gesuch"]


**Another optimisation would be to remove the 'seller' column as all the sellers are 'privat'.**

In [8]:
autos = pd.concat([autos.loc[:, :'name'], autos.loc[:, 'offer_type':]],  axis=1)
autos.info()


<class 'pandas.core.frame.DataFrame'>
Index: 49999 entries, 0 to 49999
Data columns (total 19 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   date_crawled        49999 non-null  object 
 1   name                49999 non-null  object 
 2   offer_type          49999 non-null  object 
 3   price_dollars       49999 non-null  float64
 4   abtest              49999 non-null  object 
 5   vehicle_type        44904 non-null  object 
 6   registration_year   49999 non-null  int64  
 7   gearbox             47320 non-null  object 
 8   power_ps            49999 non-null  int64  
 9   model               47241 non-null  object 
 10  odometer_km         49999 non-null  float64
 11  registration_month  49999 non-null  int64  
 12  fuel_type           45518 non-null  object 
 13  brand               49999 non-null  object 
 14  unrepaired_damage   40171 non-null  object 
 15  ad_created          49999 non-null  object 
 16  nr_of_pic

**Remove the offer_type columns as the dataset implicitly reflects the cars that are sold.**

In [9]:
autos = pd.concat([autos.loc[:, :'name'], autos.loc[:, 'price_dollars':]],  axis=1)
autos.info()

<class 'pandas.core.frame.DataFrame'>
Index: 49999 entries, 0 to 49999
Data columns (total 18 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   date_crawled        49999 non-null  object 
 1   name                49999 non-null  object 
 2   price_dollars       49999 non-null  float64
 3   abtest              49999 non-null  object 
 4   vehicle_type        44904 non-null  object 
 5   registration_year   49999 non-null  int64  
 6   gearbox             47320 non-null  object 
 7   power_ps            49999 non-null  int64  
 8   model               47241 non-null  object 
 9   odometer_km         49999 non-null  float64
 10  registration_month  49999 non-null  int64  
 11  fuel_type           45518 non-null  object 
 12  brand               49999 non-null  object 
 13  unrepaired_damage   40171 non-null  object 
 14  ad_created          49999 non-null  object 
 15  nr_of_pictures      49999 non-null  int64  
 16  postal_co

**'price_dollars' converting 0.0 values to np.nan**

   As we'll see below there are a lot of cars priced 0.0, and these values have to be converted to np.nan.

In [10]:
all_the_zeroes = autos["price_dollars"] < 500
autos.loc[all_the_zeroes,"price_dollars"] = np.nan
autos["price_dollars"].value_counts(dropna = False) 

price_dollars
NaN        4888
500.0       781
1500.0      734
2500.0      643
1000.0      639
           ... 
5049.0        1
2651.0        1
7085.0        1
2498.0        1
16995.0       1
Name: count, Length: 2219, dtype: int64

In [11]:
autos[autos["price_dollars"].isnull()]

date_crawled                                               name  \
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
8      2016-03-22 16:51:34                                         Seat_Arosa   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
...                    ...                                                ...   
49960  2016-03-25 22:51:55            Ford_KA_zu_verschenken_***Reserviert***   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49989  2016-03-11 19:50:37        VW_Polo_zum_Ausschlachten_oder_Wiederaufbau   

       price_dollars   abtest vehicle_type  registration_year  gearbox  \
6                NaN     test    limousine               1995  manuell   
8                NaN     test          NaN               2000  manuell   
11               NaN  control          NaN               2000      NaN   
16               NaN     test        kombi               1999  manuell   
18               NaN  control          bus               1995  manuell   
...              ...      ...          ...                ...      ...   
49960            NaN  control   kleinwagen               1999  manuell   
49974            NaN  control       cabrio               1983  manuell   
49980            NaN  control        kombi               1995  manuell   
49984            NaN     test          NaN               2000      NaN   
49989            NaN     test   kleinwagen               1997  manuell   

       power_ps   model  odometer_km  registration_month fuel_type  \
6            90    golf     150000.0                   8    benzin   
8             0   arosa     150000.0                  10       NaN   
11            0     NaN     150000.0                   0    benzin   
16          101  vectra     150000.0                   5    benzin   
18          150     3er     150000.0                   0    benzin   
...         ...     ...          ...                 ...       ...   
49960        60      ka     150000.0                   6    benzin   
49974        70    golf     150000.0                   2    benzin   
49980       105  escort     125000.0                   3    benzin   
49984         0     NaN     150000.0                   0       NaN   
49989         0    polo     150000.0                   5    benzin   

                brand unrepaired_damage           ad_created  nr_of_pictures  \
6          volkswagen               NaN  2016-03-20 00:00:00               0   
8                seat              nein  2016-03-22 00:00:00               0   
11      mercedes_benz               NaN  2016-03-16 00:00:00               0   
16               opel              nein  2016-03-16 00:00:00               0   
18                bmw               NaN  2016-03-26 00:00:00               0   
...               ...               ...                  ...             ...   
49960            ford               NaN  2016-03-25 00:00:00               0   
49974      volkswagen              nein  2016-03-20 00:00:00               0   
49980            ford               NaN  2016-03-12 00:00:00               0   
49984  sonstige_autos               NaN  2016-03-31 00:00:00               0   
49989      volkswagen                ja  2016-03-11 00:00:00               0   

       postal_code            last_seen  
6            31535  2016-03-23 02:48:59  
8             7426  2016-03-26 18:18:10  
11           17498  2016-03-16 18:45:34  
16           57299  2016-03-18 05:29:37  
18           54329  2016-04-02 12:16:41  
...            ...                  ...  
49960        34355  2016-03-25 22:51:55  
49974         

**'price_dollars' cleaning of the large outliers**

As the selection below shows, there are certain high outliers that are not reflecting reality, e.g.:
- A threshold which would be inclusive of a sensitive price (47634, Ferrari_FXX) is 3890000, anything greater should be investigated.  
- Again the cars that have prices greater than 500000 have to be investigated for outliers. We'll observe that appart from the Ferrari cars all the rest are banter.

In [12]:
autos[autos[ "price_dollars"]> 3890000 ] = np.nan
autos["price_dollars"].value_counts(dropna = False)

price_dollars
NaN        4896
500.0       781
1500.0      734
2500.0      643
1200.0      639
           ... 
79933.0       1
5198.0        1
18890.0       1
15470.0       1
16995.0       1
Name: count, Length: 2214, dtype: int64

In [13]:
autos[autos["price_dollars"] > 500000]

date_crawled                              name  price_dollars  \
514    2016-03-17 09:53:08  Ford_Focus_Turnier_1.6_16V_Style       999999.0   
7814   2016-04-04 11:53:31                       Ferrari_F40      1300000.0   
22947  2016-03-22 12:54:19        Bmw_530d_zum_ausschlachten      1234566.0   
37585  2016-03-29 11:38:54               Volkswagen_Jetta_GT       999990.0   
43049  2016-03-21 19:53:52                     2_VW_Busse_T3       999999.0   
47634  2016-04-04 21:25:21                       Ferrari_FXX      3890000.0   

        abtest vehicle_type  registration_year    gearbox  power_ps  \
514       test        kombi             2009.0    manuell     101.0   
7814   control        coupe             1992.0        NaN       0.0   
22947  control        kombi             1999.0  automatik     190.0   
37585     test    limousine             1985.0    manuell     111.0   
43049     test          bus             1981.0    manuell      70.0   
47634     test        coupe             2006.0        NaN     799.0   

             model  odometer_km  registration_month fuel_type           brand  \
514          focus     125000.0                 4.0    benzin            ford   
7814           NaN      50000.0                12.0       NaN  sonstige_autos   
22947          NaN     150000.0                 2.0    diesel             bmw   
37585        jetta     150000.0                12.0    benzin      volkswagen   
43049  transporter     150000.0                 1.0    benzin      volkswagen   
47634          NaN       5000.0                 7.0       NaN  sonstige_autos   

      unrepaired_damage           ad_created  nr_of_pictures  postal_code  \
514                nein  2016-03-17 00:00:00             0.0      12205.0   
7814               nein  2016-04-04 00:00:00             0.0      60598.0   
22947               NaN  2016-03-22 00:00:00             0.0      17454.0   
37585                ja  2016-03-29 00:00:00             0.0      50997.0   
43049               NaN  2016-03-21 00:00:00             0.0      99880.0   
47634              nein  2016-04-04 00:00:00             0.0      60313.0   

                 last_seen  
514    2016-04-06 07:17:35  
7814   2016-04-05 11:34:11  
22947  2016-04-02 03:17:32  
37585  2016-03-29 11:38:54  
43049  2016-03-28 17:18:28  
47634  2016-04-05 12:07:37

In [14]:
autos.loc[
    (autos["price_dollars"] > 500000) & 
    (autos["price_dollars"] < 3890000) & 
    ~(autos["name"].str.contains("Ferrari", na=False)),
    "price_dollars"
] = np.nan
print(autos[autos["price_dollars"] > 500000])

autos["price_dollars"].value_counts(dropna = False)

              date_crawled         name  price_dollars   abtest vehicle_type  \
7814   2016-04-04 11:53:31  Ferrari_F40      1300000.0  control        coupe   
47634  2016-04-04 21:25:21  Ferrari_FXX      3890000.0     test        coupe   

       registration_year gearbox  power_ps model  odometer_km  \
7814              1992.0     NaN       0.0   NaN      50000.0   
47634             2006.0     NaN     799.0   NaN       5000.0   

       registration_month fuel_type           brand unrepaired_damage  \
7814                 12.0       NaN  sonstige_autos              nein   
47634                 7.0       NaN  sonstige_autos              nein   

                ad_created  nr_of_pictures  postal_code            last_seen  
7814   2016-04-04 00:00:00             0.0      60598.0  2016-04-05 11:34:11  
47634  2016-04-04 00:00:00             0.0      60313.0  2016-04-05 12:07:37  


price_dollars
NaN        4900
500.0       781
1500.0      734
2500.0      643
1000.0      639
           ... 
5049.0        1
2651.0        1
7085.0        1
2498.0        1
16995.0       1
Name: count, Length: 2211, dtype: int64

In [15]:
autos["price_dollars"].describe()

count    4.509900e+04
mean     6.435459e+03
std      2.138513e+04
min      5.000000e+02
25%      1.500000e+03
50%      3.500000e+03
75%      7.900000e+03
max      3.890000e+06
Name: price_dollars, dtype: float64

In [16]:
autos["price_dollars"].dtypes
(autos["price_dollars"]).isna().sum()

autos["odometer_km"].describe()
autos["odometer_km"].value_counts()

odometer_km
150000.0    32418
125000.0     5170
100000.0     2168
90000.0      1757
80000.0      1436
70000.0      1230
60000.0      1164
50000.0      1026
5000.0        967
40000.0       818
30000.0       789
20000.0       784
10000.0       264
Name: count, dtype: int64

## Price and Odometer Brief Conclusion

Price Conclusions: The mean price value is ~6400$. The minimum representative value in my opinion is 500. There are some high outliers which (after cleaning) are real, e.g. the Ferrari prices. After cleaning the price_dollars column we have 4900 NaN entries. **These NaN values have to be analysed and replaced in different transaction category: "auction", "rent", "search listings". 75% of the car listings cost less than 7900 dollars.** 

Odometer Conclusions: The mean odometer_km value is ~125734 km. The minimum value is 5000km. The maximum value in the listings is 150000km. **There are 32418 values of 150000 km, which oculd mean that there is a causal correlation between hitting the 150000km mark and the decision of the owner to sell. The second most frequent odometer valus is 125000km(5170 listings) and 100000(listings).**

## Date Columns Data

   Currently there are 5 date related columns
- `date_crawled`: added by the crawler;
- `last_seen`: added by the crawler;
- `ad_created`: from the website;
- `registration_month`: from the website;
- `registration_year`: from the website;

Right now, the **date_crawled**, **last_seen**, and **ad_created** columns are all identified as **string** values by pandas. Because these three columns are represented as strings, we need to convert the data into a numerical representation so we can understand it quantitatively. The other two columns are represented as numeric values, so we can use methods like **Series.describe()** to understand the distribution without any extra data processing.

**Below we will create frequency of the date_crawled, last_seen and ad_created columns expressed as percentages.**

In [17]:
date_crawled_distribution = autos["date_crawled"].str[:10]
date_crawled_values = date_crawled_distribution.value_counts(normalize=True, dropna=False).sort_index() * 100

last_seen_distribution = autos["last_seen"].str[:10]
last_seen_values = last_seen_distribution.value_counts(normalize=True, dropna=False).sort_index() * 100

ad_created_distribution = autos["ad_created"].str[:10]
ad_created_values = ad_created_distribution.value_counts(normalize=True, dropna=False).sort_index() * 100

print(date_crawled_values, "\n",last_seen_values, "\n", ad_created_values)

date_crawled
2016-03-05    2.538051
2016-03-06    1.394028
2016-03-07    3.596072
2016-03-08    3.326067
2016-03-09    3.320066
2016-03-10    3.212064
2016-03-11    3.248065
2016-03-12    3.676074
2016-03-13    1.556031
2016-03-14    3.662073
2016-03-15    3.398068
2016-03-16    2.950059
2016-03-17    3.152063
2016-03-18    1.306026
2016-03-19    3.490070
2016-03-20    3.782076
2016-03-21    3.750075
2016-03-22    3.292066
2016-03-23    3.238065
2016-03-24    2.910058
2016-03-25    3.174063
2016-03-26    3.248065
2016-03-27    3.104062
2016-03-28    3.484070
2016-03-29    3.416068
2016-03-30    3.362067
2016-03-31    3.190064
2016-04-01    3.380068
2016-04-02    3.540071
2016-04-03    3.866077
2016-04-04    3.652073
2016-04-05    1.310026
2016-04-06    0.318006
2016-04-07    0.142003
NaN           0.016000
Name: proportion, dtype: float64 
 last_seen
2016-03-05     0.108002
2016-03-06     0.442009
2016-03-07     0.536011
2016-03-08     0.758015
2016-03-09     0.984020
2016-03-10     1.

Below we defined the threshold for the minimum value to be 1929 and for the maximum value to be 2018. Anything outside this limit will be assigned to np.nan. 

In [18]:
autos.loc[autos["registration_year"] > 2018, "registration_year"] = np.nan

autos.loc[autos["registration_year"] < 1929, "registration_year"] = np.nan


In [19]:
print(autos["registration_year"].describe(), "\n")
print(autos["registration_year"].info(), "\n")

percentage_by_regyear = autos["registration_year"].value_counts(normalize=True)*100
print(percentage_by_regyear.head(20))

count    49954.000000
mean      2003.385214
std          7.579718
min       1929.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       2018.000000
Name: registration_year, dtype: float64 

<class 'pandas.core.series.Series'>
Index: 49999 entries, 0 to 49999
Series name: registration_year
Non-Null Count  Dtype  
--------------  -----  
49954 non-null  float64
dtypes: float64(1)
memory usage: 781.2 KB
None 

registration_year
2000.0    6.714177
2005.0    6.033551
1999.0    6.003523
2004.0    5.479041
2003.0    5.459022
2006.0    5.420987
2001.0    5.408976
2002.0    5.070665
1998.0    4.910518
2007.0    4.612243
2008.0    4.466109
2009.0    4.199864
1997.0    4.059735
2011.0    3.271009
2010.0    3.196941
2017.0    2.906674
1996.0    2.890659
2012.0    2.648437
2016.0    2.634424
1995.0    2.626416
Name: proportion, dtype: float64


The majority of cars in the dataset were registered between 1998 and 2008, with a median year of 2003. There are some older cars going back to 1929 and some newer ones up to 2018, which might represent outliers or data entry errors.

The most common registration years are 2000, 2005, and 1999, each representing around 6% of the dataset. Overall, the distribution shows a clear concentration of cars from the early 2000s.

## Exploring Price by Brand

Explore the unique values of the "brand" column.

In [20]:
brands_frequency = autos["brand"].value_counts(normalize=True) 

In [25]:
if "sonstige_autos" in brands_frequency:
    brands_frequency = brands_frequency.drop("sonstige_autos").head(20)
    brands_frequency = brands_frequency[(brands_frequency*100) > 5]
#print(brands_frequency)

top_brands_counts = autos["brand"].value_counts()[:len(brands_frequency)]
top_brands_counts

brand
volkswagen       10685
opel              5460
bmw               5429
mercedes_benz     4733
audi              4283
ford              3478
Name: count, dtype: int64

Create a dictionary of price means values for the brands keys, which will depict a mean value for the top brands. 

In [49]:
brand_meanprice = {}
brands = list(top_brands_counts.index)
print(brands)

for b in brands:
    total_sum = 0
    prices_series = autos.loc[
    (autos["brand"] == b) & (autos["price_dollars"].notna()),
    "price_dollars"]
    brand_meanprice[b] = round(prices_series.mean())
print(brand_meanprice)    
    

['volkswagen', 'opel', 'bmw', 'mercedes_benz', 'audi', 'ford']
{'volkswagen': 5701, 'opel': 3349, 'bmw': 8515, 'mercedes_benz': 8670, 'audi': 9484, 'ford': 4267}


In [42]:
autos["brand"].value_counts(normalize=True)

brand
volkswagen        0.213738
opel              0.109220
bmw               0.108600
mercedes_benz     0.094677
audi              0.085675
ford              0.069573
renault           0.048089
peugeot           0.029125
fiat              0.026145
seat              0.018823
skoda             0.015723
mazda             0.015143
nissan            0.015083
smart             0.014023
citroen           0.014003
toyota            0.012342
sonstige_autos    0.010902
hyundai           0.009762
volvo             0.009122
mini              0.008482
mitsubishi        0.008121
honda             0.007981
kia               0.007121
alfa_romeo        0.006581
porsche           0.005881
suzuki            0.005861
chevrolet         0.005661
chrysler          0.003621
dacia             0.002580
daihatsu          0.002560
jeep              0.002200
subaru            0.002180
land_rover        0.001980
saab              0.001600
daewoo            0.001580
trabant           0.001560
jaguar            0.00

~21.4 percent of the sold cars are VW, ~11 percent Opel and ~11 percent bmw. 

## The Mean Milage and Mean Price Relationship

Create a dataframe from the **brand_meanprice** dictionary.

In [50]:
correlation_df = pd.DataFrame(pd.Series(brand_meanprice), columns = ["mean_price"])

Obtain the **mean_milage** data and concatenate to the correlation_df dataframe.

In [57]:
brand_meanmileage = {}


for b in brands:
    total_sum = 0
    mileage_series = autos.loc[
    (autos["brand"] == b) & (autos["odometer_km"].notna()),
    "odometer_km"]
    brand_meanmileage[b] = round(mileage_series.mean())


correlation_df["mean_mileage"] = brand_meanmileage

correlation_df

mean_price  mean_mileage
volkswagen           5701        128951
opel                 3349        129295
bmw                  8515        132522
mercedes_benz        8670        130882
audi                 9484        129644
ford                 4267        124153

The range of car mileages does not vary as much as the prices do by brand, instead all falling within 10% for the top brands. There is a slight trend to the more expensive vehicles having higher mileage, with the less expensive vehicles having lower mileage.